### PR TITLE
Support nonholonomic movement for slotcar

### PR DESCRIPTION
**Pull request by ddengster**
### Implemented feature

Support nonholonomic movement for slotcar

### Implementation description

Robots using the slotcar plugin can now attempt to follow a path made by nonholonomic robots via publishing to `/ackmann_path_requests` and adding the tag `<holonomic>0</holonomic>` to their sdf model files. An example script and level can be used to test in this PR https://github.com/open-rmf/rmf_demos/pull/70.

Currently the nonholonomic movement update code does not consider
- battery usage (not needed yet)
- turning accelerations (due to bugs with the rate of change code)

### CODE REVIEW

**Comment by luca-della-vedova** [comment]
Before I look too much into the code I can see from a quick glance that mostly new functions are used (i.e. the ackermann topic callback, an update specifically for non holonomic vehicles), and not too much of the slotcar pre-existing code.
Hence I wonder if it would be cleaner / easier to maintain (avoiding slotcar becoming even more complex) if the ackermann vehicles existed in a separate plugin, or do you think there is enough shared logic between ackermann drives and normal slotcar robots to justify the code being part of the slotcar plugin?

**Comment by ddengster** [comment]
There were some code like movement, battery usage, etc and I thought the potential of reusing them would be a good thing, however the movement code didn't turn out to be that suited. Still I think it can work out correctly. The battery code is another part that would be useful to have in the future.

Still I think we can stick to this arrangement - I don't feel a new movement and path construction function is enough for a new plugin.

File: rmf_robot_sim_ignition_plugins/src/slotcar.cpp [hunk 1]
```diff
@@ -164,7 +169,23 @@ void SlotcarPlugin::send_control_signals(EntityComponentManager& ecm,
...
+  // lin_vel_cmd->Data()[2] = velocities.second;
+
+  if (dataPtr->model_name() == "ambulance")
+  {
+    //lin_vel_cmd->Data()[0] = 1;
+    // ang_vel_cmd->Data()[2] = -M_PI;
+    // printf("v_robot %g velocities.first %g target_vels[0]: %g\n", v_robot, velocities.first, target_vels[0]);
+    // printf("w_robot %g velocities.second %g target_vels[1]: %g\n", w_robot, velocities.second, target_vels[1]);
+  }
+  // printf("lin_vel_cmd->Data()[0]: %g\n", lin_vel_cmd->Data()[0]);
```
Comment (luca-della-vedova) [inline rmf_robot_sim_ignition_plugins/src/slotcar.cpp:188]: Remove all of this

File: rmf_robot_sim_ignition_plugins/src/slotcar.cpp [hunk 2]
```diff
@@ -345,14 +367,42 @@ void SlotcarPlugin::PreUpdate(const UpdateInfo& info,
...
 
-  auto pose = ecm.Component<components::Pose>(_entity)->Data();
-  auto obstacle_positions = get_obstacle_positions(ecm);
+  if (!dataPtr->is_holonomic())
+  {
+    if (_remove_world_pose_cmd)
+    {
+      ecm.RemoveComponent<components::WorldPoseCmd>(_entity);
+      _remove_world_pose_cmd = false;
+    }
```
Comment (luca-della-vedova) [inline rmf_robot_sim_ignition_plugins/src/slotcar.cpp:376]: This can also be removed right?

File: rmf_robot_sim_ignition_plugins/src/slotcar.cpp [hunk 3]
```diff
@@ -345,14 +367,42 @@ void SlotcarPlugin::PreUpdate(const UpdateInfo& info,
...
+
+    auto isometry_pose = rmf_plugins_utils::convert_pose(pose);
+    auto velocities = dataPtr->update_nonholonomic(isometry_pose, dt);
+
+    //convert back to account for flips
+    pose = rmf_plugins_utils::convert_to_pose<ignition::math::Pose3d>(
+      isometry_pose);
+
+    send_control_signals(ecm, velocities, _payloads, dt);
+    return;
```
Comment (luca-della-vedova) [inline rmf_robot_sim_ignition_plugins/src/slotcar.cpp:389]: Nit, this is not needed right?

File: rmf_robot_sim_ignition_plugins/src/slotcar.cpp [hunk 4]
```diff
@@ -11,6 +11,7 @@
 #include <ignition/gazebo/components/Pose.hh>
 #include <ignition/gazebo/components/Static.hh>
 #include <ignition/gazebo/components/AxisAlignedBox.hh>
+#include <ignition/gazebo/components/PoseCmd.hh>
```
Comment (luca-della-vedova) [inline rmf_robot_sim_ignition_plugins/src/slotcar.cpp:14]: It seems that PoseCmd components are not currently used so this can be probably removed?

File: rmf_robot_sim_ignition_plugins/src/slotcar.cpp [hunk 5]
```diff
@@ -345,14 +367,42 @@ void SlotcarPlugin::PreUpdate(const UpdateInfo& info,
...
+  if (!dataPtr->is_holonomic())
+  {
+    if (_remove_world_pose_cmd)
+    {
+      ecm.RemoveComponent<components::WorldPoseCmd>(_entity);
+      _remove_world_pose_cmd = false;
+    }
+
+    auto& pose = ecm.Component<components::Pose>(_entity)->Data();
+    std::vector<Eigen::Vector3d> obstacle_positions;
```
Comment (luca-della-vedova) [inline rmf_robot_sim_ignition_plugins/src/slotcar.cpp:379]: This variable is not being used

File: rmf_robot_sim_common/src/slotcar_common.cpp [hunk 6]
```diff
@@ -243,6 +249,129 @@ void SlotcarCommon::path_request_cb(
   }
 }
 
+void SlotcarCommon::ackmann_path_request_cb(
```
Comment (luca-della-vedova) [inline rmf_robot_sim_common/src/slotcar_common.cpp:252]: I think it would be helpful for future maintenance to document the steps of this function a bit (or break it down into subfunctions), I personally find it a bit hard to follow what is happening.

File: rmf_robot_sim_common/src/slotcar_common.cpp [hunk 7]
```diff
@@ -538,6 +673,113 @@ std::pair<double, double> SlotcarCommon::update(const Eigen::Isometry3d& pose,
   return velocities;
 }
 
+std::pair<double, double> SlotcarCommon::update_nonholonomic(
```
Comment (luca-della-vedova) [inline rmf_robot_sim_common/src/slotcar_common.cpp:676]: Same as the callback I think it would be useful to maybe put a short comment before chunks of code to explain what they do, or break it down a bit.

**Review by luca-della-vedova — CHANGES_REQUESTED** [review]
Finished a first quick round of review, I think there is a bit of cleanup (there is quite a bit of commented code / unused variables scattered around), as well as documentation to help follow the logic better that can be done.

File: rmf_robot_sim_common/src/slotcar_common.cpp [hunk 8]
```diff
@@ -243,6 +249,129 @@ void SlotcarCommon::path_request_cb(
   }
 }
 
+void SlotcarCommon::ackmann_path_request_cb(
+  const rmf_fleet_msgs::msg::PathRequest::SharedPtr msg)
+{
+  if (path_request_valid(msg) == false)
+    return;
+  if (model_name() != "ambulance") // ambulance only
+    return;
```
[the remark this answers is not in the document]
Comment (ddengster) [reply rmf_robot_sim_common/src/slotcar_common.cpp]: account for robot names 20a8e48

File: rmf_robot_sim_common/src/slotcar_common.cpp [hunk 9]
```diff
@@ -243,6 +249,129 @@ void SlotcarCommon::path_request_cb(
...
+    double half_turn_arc = M_PI / 2.0 - half_bend_delta; // right angle tri, 90 - half_turn_delta
+
+    // the computation for min_turning_radius using drive speed and turn speed
+    // reference:
+    // https://www.vboxautomotive.co.uk/downloads/Calculating%20Radius%20of%20Turn%20from%20Yaw.pdf
+    //double min_turning_radius = 0.5;
+    double min_turning_radius =
+      (this->_nominal_drive_speed * 0.2777 / this->_nominal_turn_speed *
+      0.0174);
+    //printf("min_turning_radius: %g\n", min_turning_radius);
```
[the remark this answers is not in the document]
Comment (ddengster) [reply rmf_robot_sim_common/src/slotcar_common.cpp]: reworked mini_turning_radius to retrieve values from xml, else default to formula calculation. also removed print 20a8e48

File: rmf_robot_sim_common/src/slotcar_common.cpp [hunk 10]
```diff
@@ -266,11 +395,17 @@ std::array<double, 2> SlotcarCommon::calculate_control_signals(
...
       w_robot,
       _nominal_turn_speed,
       _nominal_turn_acceleration, _max_turn_acceleration, dt);
 
+  if (!this->_is_holonomic)
+  {
+    // @note: temporary hack to resolve weird turns
+    w_target = velocities.second;
+  }
+
```
[the remark this answers is not in the document]
Comment (ddengster) [reply rmf_robot_sim_common/src/slotcar_common.cpp:408]: this is required. turning hasn't been working correctly (something about bugs with the rate of change). I'm fully expecting to re-look at this code but for now this will have to do

File: rmf_robot_sim_ignition_plugins/src/slotcar.cpp (hunk 2 above)
Comment (ddengster) [reply rmf_robot_sim_ignition_plugins/src/slotcar.cpp]: removed 20a8e48

File: rmf_robot_sim_ignition_plugins/src/slotcar.cpp (hunk 5 above)
Comment (ddengster) [reply rmf_robot_sim_ignition_plugins/src/slotcar.cpp]: remove 20a8e48

File: rmf_robot_sim_ignition_plugins/src/slotcar.cpp (hunk 4 above)
Comment (ddengster) [reply rmf_robot_sim_ignition_plugins/src/slotcar.cpp]: removed a549f8f

File: rmf_robot_sim_ignition_plugins/src/slotcar.cpp (hunk 3 above)
Comment (ddengster) [reply rmf_robot_sim_ignition_plugins/src/slotcar.cpp]: removed return a549f8f

File: rmf_robot_sim_ignition_plugins/src/slotcar.cpp (hunk 1 above)
Comment (ddengster) [reply rmf_robot_sim_ignition_plugins/src/slotcar.cpp]: removed a549f8f

File: rmf_robot_sim_common/src/slotcar_common.cpp [hunk 11]
```diff
@@ -538,6 +673,113 @@ std::pair<double, double> SlotcarCommon::update(const Eigen::Isometry3d& pose,
   return velocities;
 }
 
+std::pair<double, double> SlotcarCommon::update_nonholonomic(
+  Eigen::Isometry3d& pose,
+  const double time)
```
[the remark this answers is not in the document]
Comment (ddengster) [reply rmf_robot_sim_common/src/slotcar_common.cpp]: removed a549f8f

File: rmf_robot_sim_common/src/slotcar_common.cpp [hunk 12]
```diff
@@ -149,6 +149,12 @@ void SlotcarCommon::init_ros_node(const rclcpp::Node::SharedPtr node)
     "/robot_mode_requests",
     10,
     std::bind(&SlotcarCommon::mode_request_cb, this, std::placeholders::_1));
+
+  _traj_sub = _ros_node->create_subscription<rmf_fleet_msgs::msg::PathRequest>(
+    "/ackmann_path_requests",
+    10,
+    std::bind(&SlotcarCommon::ackmann_path_request_cb, this,
+    std::placeholders::_1));
```
[the remark this answers is not in the document]
Comment (ddengster) [reply rmf_robot_sim_common/src/slotcar_common.cpp]: added nonholonomic check a549f8f

File: rmf_robot_sim_common/src/slotcar_common.cpp [hunk 13]
```diff
@@ -243,6 +249,129 @@ void SlotcarCommon::path_request_cb(
...
+    }
+    else
+    {
+      NonHolonomicTrajectory& last_traj = this->nonholonomic_trajectory.back();
+
+      // bend, build an intermediate spline using turn rate.
+      Eigen::Vector2d tangent0 = wp[1] + tangent_length * wp1_to_wp0_norm;
+      Eigen::Vector2d tangent1 = wp[1] + tangent_length * wp1_to_wp2_norm;
+
+      // shorten the last trajectory and set it's heading
```
[the remark this answers is not in the document]
Comment (ddengster) [reply rmf_robot_sim_common/src/slotcar_common.cpp:339]: what's the typo? `last`?

File: rmf_robot_sim_common/src/slotcar_common.cpp (hunk 6 above)
Comment (ddengster) [reply rmf_robot_sim_common/src/slotcar_common.cpp:255]: yea there's a bit of math/geometry background needed here, add comments 1ec3bde

File: rmf_robot_sim_common/src/slotcar_common.cpp (hunk 7 above)
Comment (ddengster) [reply rmf_robot_sim_common/src/slotcar_common.cpp:688]: add comments 1ec3bde

File: rmf_robot_sim_common/src/slotcar_common.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (luca-della-vedova) [reply rmf_robot_sim_common/src/slotcar_common.cpp:349]: it's -> its

**Review by luca-della-vedova — APPROVED** [review]
